File: check_entropy.m
% CHECK_ENTROPY  Check the entropy estimator against cases with known answers.
%
% Nothing here involves the neuron: these are synthetic spike trains chosen so
% that the total entropy, the noise entropy or the information is known exactly
% in advance, which is the only way to tell an estimator bug from a real
% result. Mirrors python/check_entropy.py.
%
%   check_entropy        % a few seconds

rng(0);

t_bin = 1;                    % [ms]
p     = 0.05;                 % spike probability per bin -> 50 Hz
ok    = true(1,4);

% --- 1. binarisation ------------------------------------------------------
fprintf('1. spike_bits: bin k covers ((k-1)*t_bin, k*t_bin]\n');
st = [0.4 1.0 1.6 2.4 9.9 10.0];
[bits, merged] = spike_bits(st, 10, 1);
want  = logical([1 1 1 0 0 0 0 0 0 1]);
ok(1) = isequal(bits, want) && merged == 2;
fprintf('   spikes -> [%s], %d merged   [%s]\n', ...
        num2str(double(bits)), merged, pass(ok(1)));
fprintf('   (0.4 and 1.0 share bin 1; 9.9 and 10.0 share bin 10)\n\n');

% --- 2. independent bins: entropy rate is H(p)/t_bin at every L -----------
% Compared against the entropy of the *empirical* p, which separates the
% estimator's error from the sampling error of p itself.
fprintf('2. i.i.d. Bernoulli bins: entropy rate is H(p)/t_bin, independent of L\n');
bits  = rand(1, 400000) < p;
p_hat = mean(bits);
truth = -(p_hat*log2(p_hat) + (1-p_hat)*log2(1-p_hat)) / (t_bin/1000);
L     = [1 2 4 8 12 16];
S     = spike_entropy(bits, t_bin, L);
fprintf('   true rate %.2f bits/s (empirical p = %.5f)\n', truth, p_hat);
fprintf('   L         :%s\n', sprintf('%9d', L));
fprintf('   naive  err:%s\n', sprintf('%+9.3f', S.H_total_naive - truth));
fprintf('   extrap err:%s\n', sprintf('%+9.3f', S.H_total - truth));
worst = max(abs(S.H_total - truth));
ok(2) = worst < 0.005*truth;                       % 0.5% of the true rate
fprintf('   worst %.3f bits/s out of %.0f, tolerance 0.5%%   [%s]\n\n', ...
        worst, truth, pass(ok(2)));

% --- 3. a period-P train has exactly P distinct words once L >= P-1 -------
fprintf('3. period-7 train: exactly 7 distinct words, so log2(7) bits/word\n');
bits = false(1, 21000);
bits(1:7:end) = true;
L = [2 4 6 8 12];
S = spike_entropy(bits, t_bin, L, struct('extrapolate', false));
per_word = S.H_total_naive .* S.T;
fprintf('   L        :%s\n', sprintf('%9d', L));
fprintf('   bits/word:%s\n', sprintf('%9.4f', per_word));
ok(3) = all(abs(per_word(L >= 6) - log2(7)) < 1e-6);
fprintf('   log2(7) = %.4f for L >= 6   [%s]\n\n', log2(7), pass(ok(3)));

% --- 4. the two limits of the noise entropy ------------------------------
% The first is exact and is asserted. The second is only exact for an infinite
% number of trials, so what it really shows is the size of the finite-trial
% bias -- quantified in check 5.
fprintf('4. the two limits of the noise entropy\n');
L   = [2 4 8 12];
one = rand(1, 20000) < p;

S     = spike_entropy(repmat(one, 20, 1), t_bin, L);
ok(4) = max(abs(S.H_noise)) == 0;
fprintf(['   20 identical trials -> noise entropy %.3g bits/s, ' ...
         'information = total   [%s]\n'], max(abs(S.H_noise)), pass(ok(4)));

S = spike_entropy(rand(256, 20000) < p, t_bin, L);
fprintf('   256 independent trials -> true information is 0; what is left\n');
fprintf('   is the finite-trial bias, and it grows with the word length:\n');
fprintf('   L        :%s\n', sprintf('%9d', L));
fprintf('   info     :%s   (total %.0f bits/s)\n\n', ...
        sprintf('%9.1f', S.info), max(S.H_total));

% --- 5. how many trials the noise entropy needs --------------------------
% Descriptive, not pass/fail: there is no threshold here that would not be
% arbitrary. The point is the scaling -- the bias falls roughly as 1/n_trials.
fprintf('5. finite-trial bias: residual information where the true value is 0\n');
bits = rand(512, 20000) < p;
fprintf('   trials |%s\n', sprintf('     L=%-4d', L));
for n = [8 16 32 64 128 256 512]
    S = spike_entropy(bits(1:n,:), t_bin, L);
    fprintf('   %6d |%s\n', n, sprintf('%10.1f', S.info));
end

% --- 6. the shift control sees that floor without knowing the answer ------
fprintf('\n6. the shift control recovers that floor from time-locked data\n');
one = rand(1, 20000) < p;
for n = [16 128]
    % Every trial locked to the same train, plus independent extra spikes.
    b = repmat(one, n, 1) | (rand(n, 20000) < 0.01);
    S = spike_entropy(b, t_bin, L);
    C = spike_entropy(shift_trials(b), t_bin, L);
    fprintf('   n=%-4d info  :%s\n', n, sprintf('%10.1f', S.info));
    fprintf('          floor :%s\n', sprintf('%10.1f', C.info));
end

fprintf('\n%s\n', ternary(all(ok), 'all checks passed', 'SOME CHECKS FAILED'));

% -----------------------------------------------------------------------
function s = pass(tf)
s = ternary(tf, 'ok', 'FAILED');
end

function out = ternary(tf, a, b)
if tf, out = a; else, out = b; end
end

File: python/check_entropy.py
"""Check the entropy estimator against cases whose answer is known.

Mirrors check_entropy.m.

    python3 python/check_entropy.py        # a few seconds

Nothing here involves the neuron: these are synthetic spike trains chosen so
that the total entropy, the noise entropy, or the information is known exactly
in advance, which is the only way to tell an estimator bug from a real result.
"""

import os
import sys

import numpy as np

sys.path.insert(0, os.path.dirname(os.path.abspath(__file__)))
from lif import shift_trials, spike_bits, spike_entropy   # noqa: E402

T_BIN = 1.0                         # [ms]
P = 0.05                            # spike probability per bin -> 50 Hz


def H_bernoulli(p):
    """Entropy of one independent bin, in bits."""
    return -(p * np.log2(p) + (1 - p) * np.log2(1 - p))


def check_binarisation():
    print("1. spike_bits: bin k covers ((k-1)*t_bin, k*t_bin]")
    st = np.array([0.4, 1.0, 1.6, 2.4, 9.9, 10.0])
    bits, merged = spike_bits([st], 10.0, 1.0)
    want = np.array([1, 1, 1, 0, 0, 0, 0, 0, 0, 1])
    ok = np.array_equal(bits[0], want) and merged == 2
    print(f"   spikes {st} -> {bits[0]}, {merged} merged   [{'ok' if ok else 'FAILED'}]")
    print("   (0.4 and 1.0 share bin 1; 9.9 and 10.0 share bin 10)\n")
    return ok


def check_iid(rng):
    """Independent bins: the entropy rate is H(p)/t_bin at every word length.

    Compared against the entropy of the *empirical* p, which separates the
    estimator's error from the sampling error of p itself.
    """
    print("2. i.i.d. Bernoulli bins: entropy rate is H(p)/t_bin, independent of L")
    bits = (rng.random((1, 400_000)) < P).astype(np.uint8)
    truth = H_bernoulli(bits.mean()) / (T_BIN / 1000)     # empirical p
    L = np.array([1, 2, 4, 8, 12, 16])
    S = spike_entropy(bits, T_BIN, L)
    print(f"   true rate {truth:.2f} bits/s (empirical p = {bits.mean():.5f})")
    print("   L         :" + "".join(f"{v:9d}" for v in L))
    print("   naive  err:" + "".join(f"{v:+9.3f}" for v in S.H_total_naive - truth))
    print("   extrap err:" + "".join(f"{v:+9.3f}" for v in S.H_total - truth))
    worst = np.max(np.abs(S.H_total - truth))
    ok = worst < 0.005 * truth                     # 0.5% of the true rate
    print(f"   worst {worst:.3f} bits/s out of {truth:.0f}, tolerance 0.5%"
          f"   [{'ok' if ok else 'FAILED'}]\n")
    return ok


def check_periodic():
    """A period-p train has exactly p distinct words once L >= p-1."""
    print("3. period-7 train: exactly 7 distinct words, so log2(7) bits/word")
    bits = np.zeros((1, 21_000), dtype=np.uint8)
    bits[0, ::7] = 1
    L = np.array([2, 4, 6, 8, 12])
    S = spike_entropy(bits, T_BIN, L, extrapolate=False)
    per_word = S.H_total_naive * S.T
    print("   L        :" + "".join(f"{v:9d}" for v in L))
    print("   bits/word:" + "".join(f"{v:9.4f}" for v in per_word))
    ok = np.allclose(per_word[L >= 6], np.log2(7), atol=1e-6)
    print(f"   log2(7) = {np.log2(7):.4f} for L >= 6   [{'ok' if ok else 'FAILED'}]\n")
    return ok


def check_noise_entropy(rng):
    """Two limits: identical trials carry no noise, independent trials carry
    no information.

    The first is exact and is asserted. The second is only exact for an
    infinite number of trials, so what it really shows is the size of the
    finite-trial bias -- quantified in check 5.
    """
    print("4. the two limits of the noise entropy")
    L = np.array([2, 4, 8, 12])
    one = (rng.random((1, 20_000)) < P).astype(np.uint8)

    S = spike_entropy(np.repeat(one, 20, axis=0), T_BIN, L)
    max_noise = np.nanmax(np.abs(S.H_noise))
    ok_a = max_noise == 0
    print(f"   20 identical trials -> noise entropy {max_noise:.3g} bits/s, "
          f"information = total   [{'ok' if ok_a else 'FAILED'}]")

    bits = (rng.random((256, 20_000)) < P).astype(np.uint8)
    S = spike_entropy(bits, T_BIN, L)
    print("   256 independent trials -> true information is 0; what is left")
    print("   is the finite-trial bias, and it grows with the word length:")
    print("   L        :" + "".join(f"{v:9d}" for v in L))
    print(f"   info     :" + "".join(f"{v:9.1f}" for v in S.info)
          + f"   (total {S.H_total.max():.0f} bits/s)")
    print()
    return ok_a


def check_trial_count(rng):
    """How many trials the noise entropy needs, and whether the shift control
    sees the shortfall coming.

    Descriptive, not pass/fail: there is no threshold here that would not be
    arbitrary. The point is the scaling -- the bias falls roughly as 1/n_trials
    -- and that the control tracks it without knowing the answer.
    """
    print("5. finite-trial bias: residual information where the true value is 0")
    L = np.array([4, 8, 12, 16])
    bits = (rng.random((512, 20_000)) < P).astype(np.uint8)
    print("   trials |" + "".join(f"     L={v:<4d}" for v in L))
    for n in (8, 16, 32, 64, 128, 256, 512):
        S = spike_entropy(bits[:n], T_BIN, L)
        print(f"   {n:6d} |" + "".join(f"{v:10.1f}" for v in S.info))

    print("\n6. the shift control recovers that floor from time-locked data")
    one = (rng.random((1, 20_000)) < P).astype(np.uint8)
    for n in (16, 128):
        # Every trial locked to the same train, plus independent extra spikes.
        b = np.maximum(np.repeat(one, n, axis=0),
                       (rng.random((n, 20_000)) < 0.01).astype(np.uint8))
        S = spike_entropy(b, T_BIN, L)
        C = spike_entropy(shift_trials(b, rng), T_BIN, L)
        print(f"   n={n:<4d} info  :" + "".join(f"{v:10.1f}" for v in S.info))
        print("          floor :" + "".join(f"{v:10.1f}" for v in C.info))
    print()
    return True


if __name__ == "__main__":
    rng = np.random.default_rng(0)
    results = [check_binarisation(), check_iid(rng), check_periodic(),
               check_noise_entropy(rng), check_trial_count(rng)]
    print("all checks passed" if all(results) else "SOME CHECKS FAILED")
